### PR TITLE
Watchdogs

### PR DESCRIPTION
**Pull request by Benichiwa**
This implements the watchdog feature as mentioned in the discussion: "Watchdogs #1379". It may be causing issues with some single-threaded programs.

### CODE REVIEW

**Comment by Benichiwa** [comment]
CI / unit-tests / run is failing as a result of a watchdog parsing error in "Watchdog.lf" during the roundTripTests(). Although this error gets thrown while running the tests on my local machine, this is not the case when running `lfc Watchdog.lf`.

File: org.lflang/src/org/lflang/ASTUtils.java [hunk 1]
```diff
@@ -1059,6 +1075,11 @@ public static InferredType getInferredType(Port p) {
         return getInferredType(p.getType(), null);
     }
 
+    //FIXME: modif4watchdogs
```
Comment (lhstrh) [inline org.lflang/src/org/lflang/ASTUtils.java:1078]: According to the grammar, watchdogs don't have a value or type, so there appears to be nothing actionable here. Remove?

File: org.lflang/src/org/lflang/LinguaFranca.xtext [hunk 2]
```diff
@@ -91,6 +91,7 @@ Reactor:
         | (connections+=Connection)
         | (reactions+=Reaction)
         | (modes+=Mode)
+        | (watchdogs+=Watchdog)
```
Comment (lhstrh) [inline org.lflang/src/org/lflang/LinguaFranca.xtext:94]: Can you move this up, right after the actions?

File: org.lflang/src/org/lflang/LinguaFranca.xtext [hunk 3]
```diff
@@ -176,7 +177,8 @@ Mode:
         (actions+=Action) |
         (instantiations+=Instantiation) |
         (connections+=Connection) |
-        (reactions+=Reaction)
+        (reactions+=Reaction) |
+        (watchdogs+=Watchdog)
```
Comment (lhstrh) [inline org.lflang/src/org/lflang/LinguaFranca.xtext:181]: Idem (watchdogs should come after actions)

File: org.lflang/src/org/lflang/ast/IsEqual.java [hunk 4]
```diff
@@ -487,6 +491,17 @@ public Boolean caseArraySpec(ArraySpec object) {
             .conclusion;
     }
 
+    //FIXME: modif4watchdogs
```
Comment (lhstrh) [inline org.lflang/src/org/lflang/ast/IsEqual.java:494]:
```suggestion
```

File: org.lflang/src/org/lflang/ast/IsEqual.java [hunk 5]
```diff
@@ -392,12 +394,14 @@ public Boolean caseElement(Element object) {
...
     @Override
     public Boolean caseTypedVariable(TypedVariable object) {
+        //FIXME: modif4watchdogs
         throw thereIsAMoreSpecificCase(TypedVariable.class, Port.class, Action.class);
     }
 
     @Override
     public Boolean caseVariable(Variable object) {
-        throw thereIsAMoreSpecificCase(Variable.class, TypedVariable.class, Timer.class, Mode.class);
+        //FIXME: modif4watchdogs
```
Comment (lhstrh) [inline org.lflang/src/org/lflang/ast/IsEqual.java:403]:
```suggestion
```

File: org.lflang/src/org/lflang/ast/ToLf.java [hunk 6]
```diff
@@ -653,6 +656,41 @@ public MalleableString caseDeadline(Deadline object) {
     return handler(object, "deadline", Deadline::getDelay, Deadline::getCode);
   }
 
+  //FIXME: modif4watchdogs
```
Comment (lhstrh) [inline org.lflang/src/org/lflang/ast/ToLf.java:659]:
```suggestion
```

**Review by lhstrh — CHANGES_REQUESTED** [review]
I'll continue my review once all the FIXMEs are gone...

**Pull request closed without merging**